### DISCUSSION TITLE
InterfaceKernel for ArrayVariable

### DISCUSSION THREAD

**Discussion by schakrabortygithub** [idaholab/moose#26782] [Q&A General] [answered]
Hi,

This is a general question regarding mooseframework. Not a bug or an issue.
The problem that I am working on needs to develop an InterfaceKernel for ArrayVariable, because the model has many solution variables. Those variables can be grouped in few of the ArrayVariables. 

During implementation I observed that,

In most of the base classes ([KernelBase.h](https://github.com/idaholab/moose/blob/next/framework/include/kernels/KernelBase.h), [DGKernelBase.h](https://github.com/idaholab/moose/blob/next/framework/include/dgkernels/DGKernelBase.h), [NodalBCBase.h](https://github.com/idaholab/moose/blob/next/framework/include/bcs/NodalBCBase.h) ) resudual and jacobian related functions at QP (e.g., computeQpResidual, computeQpJacobian, computeQpOffDiagJacobian) are not declared in their corresponding base classes. Rather they are declared in the next level of derived classes ([Kernel.h](https://github.com/idaholab/moose/blob/next/framework/include/kernels/Kernel.h). [ArrayKernel.h](https://github.com/idaholab/moose/blob/next/framework/include/kernels/ArrayKernel.h),    [DGKernel.h](https://github.com/idaholab/moose/blob/next/framework/include/dgkernels/DGKernel.h), [ArrayDGKernel.h](https://github.com/idaholab/moose/blob/next/framework/include/dgkernels/ArrayDGKernel.h)  etc). This is good because those classes need to inherit common stuff(element, neighbor, QP etc) from their base class([kernel, Arraykernel]<= KernelBase, [DGKernel, ArrayDGKernel]<=DGKernelBase etc.). And computeQpResidual, computeQpJacobian, computeQpOffDiagJacobian can have different return types based on the variable type(real, array) they are operating on.

**One exception is in the InterfaceKernelBase.h**. Where computeQpJacobian, computeQpOffDiagJacobian are declared in the base class itself, with a return type 'Real'. I was wondering if there is any reason for doing that?

Because, when I write the class ArrayInterfaceKernel.h derived from  InterfaceKernelBase.h for ArrayVariable, return type mismatch for computeQpJacobian and computeQpOffDiagJacobian prohibits to inherit(and override) those two functions from that base class.

https://github.com/idaholab/moose/blob/27c9045c79dad1c7081f8f13a0d917df4a3fbddc/framework/include/interfacekernels/InterfaceKernelBase.h#L46-L54

Best
Subhendu Chakraborty
Los Alamos National Laboratory

**Comment by YaqiWang** [comment, marked as answer]
I did not check the code, but we probably did not think of array variables when developing `InterfaceKernelBase`. Thanks for working on this! I created an issue https://github.com/idaholab/moose/issues/26268 for this at the end of last year for your information.

**Reply by GiudGiud** [reply]
This sounds right. Array variables were a later addition and interface kernels are also a rarer need
